squirrel/static_analyser/analyser.cpp

```diff
@@ -1,62 +1,3581 @@
+  checkSameOperands(expr);
+  checkAlwaysTrueOrFalse(expr);
+  checkDeclarationInArith(expr);
+  checkIntDivByZero(expr);
+  checkPotentiallyNullableOperands(expr);
+  checkBitwiseToBool(expr);
+  checkCompareWithBool(expr);
+  checkCopyOfExpression(expr);
+  checkConstInBoolExpr(expr);
+  checkShiftPriority(expr);
+  checkCanReturnNull(expr);
+  checkCompareWithContainer(expr);
+  checkBoolToStrangePosition(expr);
+  checkNewSlotNameMatch(expr);
+  checkPlusString(expr);
+
+  switch (expr->op())
+  {
+  case TO_NULLC:
+    visitBinaryBranches(expr, &nullValue, RT_NULL, 0);
+    break;
+  case TO_ANDAND:
+    visitBinaryBranches(expr, &trueValue, 0, RT_NULL);
+    break;
+  case TO_OROR:
+    visitBinaryBranches(expr, &falseValue, RT_NULL, 0);
+    break;
+  case TO_PLUSEQ:
+  case TO_MINUSEQ:
+  case TO_MULEQ:
+  case TO_DIVEQ:
+  case TO_MODEQ:
+  case TO_NEWSLOT:
+  case TO_ASSIGN:
+  case TO_INEXPR_ASSIGN:
+    expr->rhs()->visit(this);
+    expr->lhs()->visit(this);
+    break;
+  default:
+    Visitor::visitBinExpr(expr);
+    break;
+  }
+
+  applyBinaryToScope(expr);
+}
+
+void CheckerVisitor::visitTerExpr(TerExpr *expr) {
+  checkTernaryPriority(expr);
+  checkSameValues(expr);
+  checkAlwaysTrueOrFalse(expr);
+
+  expr->a()->visit(this);
+  VarScope *trunkScope = currentScope;
+
+  VarScope *ifTrueScope = trunkScope->copy(arena);
+  currentScope = ifTrueScope;
+  setExpression(expr->a(), &trueValue, 0, RT_NULL);
+  expr->b()->visit(this);
+
+  VarScope *ifFalseScope = trunkScope->copy(arena);
+  currentScope = ifFalseScope;
+  setExpression(expr->a(), &falseValue, RT_NULL, 0);
+  expr->c()->visit(this);
+
+  trunkScope->merge(ifTrueScope);
+  trunkScope->merge(ifFalseScope);
+
+  ifTrueScope->~VarScope();
+  ifFalseScope->~VarScope();
+
+  currentScope = trunkScope;
+}
+
+void CheckerVisitor::visitCallExpr(CallExpr *expr) {
+
+
+  checkExtendToAppend(expr);
+  checkAlreadyRequired(expr);
+  checkCallNullable(expr);
+
+  applyCallToScope(expr);
+
+  Visitor::visitCallExpr(expr);
+}
+
+void CheckerVisitor::checkBoolIndex(const GetTableExpr *expr) {
+  const Expr *key = maybeEval(expr->key())->asExpression();
+
+  if (isBooleanResultOperator(key->op())) {
+    report(expr->key(), DiagnosticsId::DI_BOOL_AS_INDEX);
+  }
+}
+
+void CheckerVisitor::checkNullableIndex(const GetTableExpr *expr) {
+  const Expr *key = maybeEval(expr->key());
+
+  if (isPotentiallyNullable(key)) {
+    report(expr->key(), DiagnosticsId::DI_POTENTILLY_NULLABLE_INDEX);
+  }
+}
+
+void CheckerVisitor::visitGetFieldExpr(GetFieldExpr *expr) {
+  checkAccessNullable(expr);
+
+  Visitor::visitGetFieldExpr(expr);
+}
+
+void CheckerVisitor::visitGetTableExpr(GetTableExpr *expr) {
+  checkBoolIndex(expr);
+  checkNullableIndex(expr);
+  checkAccessNullable(expr);
+
+  Visitor::visitGetTableExpr(expr);
+}
+
+void CheckerVisitor::checkDuplicateSwitchCases(SwitchStatement *swtch) {
+  const auto &cases = swtch->cases();
+
+  for (int32_t i = 0; i < cases.size() - 1; ++i) {
```

Comment:
will out of range on empty switch operator, because of unsigned .size()

Reply:
done